src/proc/validator.rs

```diff
@@ -28,8 +81,242 @@ impl Validator {
+                            | (crate::ShaderStage::Fragment, crate::BuiltIn::SampleIndex)
+                            | (crate::ShaderStage::Compute, crate::BuiltIn::GlobalInvocationId)
+                            | (crate::ShaderStage::Compute, crate::BuiltIn::LocalInvocationId)
+                            | (crate::ShaderStage::Compute, crate::BuiltIn::LocalInvocationIndex)
+                            | (crate::ShaderStage::Compute, crate::BuiltIn::WorkGroupId) => 0,
+                            _ => return Err(EntryPointError::InvalidBuiltIn(built_in)),
+                        },
+                        Some(crate::Binding::Location(loc)) => 1 << loc,
+                        Some(crate::Binding::Resource { .. }) => unreachable!(),
+                        None => 0,
+                    };
+                    if location_in_mask & mask != 0 {
+                        return Err(EntryPointError::BindingCollision(var_handle));
+                    }
+                    location_in_mask |= mask;
+                    crate::GlobalUse::LOAD
+                }
+                crate::StorageClass::Output => {
+                    let mask = match var.binding {
+                        Some(crate::Binding::BuiltIn(built_in)) => match (stage, built_in) {
+                            (crate::ShaderStage::Vertex, crate::BuiltIn::Position)
+                            | (crate::ShaderStage::Vertex, crate::BuiltIn::PointSize)
+                            | (crate::ShaderStage::Vertex, crate::BuiltIn::ClipDistance)
+                            | (crate::ShaderStage::Fragment, crate::BuiltIn::FragDepth) => 0,
+                            _ => return Err(EntryPointError::InvalidBuiltIn(built_in)),
+                        },
+                        Some(crate::Binding::Location(loc)) => 1 << loc,
+                        Some(crate::Binding::Resource { .. }) => unreachable!(),
+                        None => 0,
+                    };
+                    if location_out_mask & mask != 0 {
+                        return Err(EntryPointError::BindingCollision(var_handle));
+                    }
+                    location_out_mask |= mask;
+                    crate::GlobalUse::STORE
+                }
+                crate::StorageClass::Constant => crate::GlobalUse::LOAD,
+                crate::StorageClass::Uniform | crate::StorageClass::StorageBuffer => {
+                    //TODO: built-in checks?
+                    let mut storage_usage = crate::GlobalUse::empty();
+                    if var.storage_access.contains(crate::StorageAccess::LOAD) {
+                        storage_usage |= crate::GlobalUse::LOAD;
+                    }
+                    if var.storage_access.contains(crate::StorageAccess::STORE) {
+                        storage_usage |= crate::GlobalUse::STORE;
+                    }
+                    //TODO: make this more explicit somehow
+                    if storage_usage.is_empty() {
```

Comment:
I don't really understand this statement. What needs to happen when the user uses an Uniform or StorageBuffer storage class? Don't we just need to explicitly state that this is a syntax error?

Or is it possible to not have a storage usage specified beforehand?

Reply:
I'll remove the todo for now. Technically, the storage access flags being non-zero means that it's a storage, so IR is actually fine.